File: flutter/lib/domain/private_agent/orchestrator.dart
import 'dart:async';

import '../../data/services/device_control_service.dart';
import '../../data/services/mobile_use_action_handler.dart';
import '../../data/services/mobile_use_ai_service.dart';
import 'executor.dart';
import 'types.dart';

/// The public `MobileUseAgent` interface that Beatrice uses to talk to
/// PrivateAgent.
///
/// It receives a structured task, validates it, runs the executor loop,
/// streams progress events, honours cancellation, and returns a structured
/// result.
class MobileUseAgent {
  final TaskExecutor _executor;

  bool _cancelRequested = false;
  TaskStatus _currentStatus = TaskStatus.planning;
  String _currentMessage = '';
  int _currentStep = 0;
  int _currentMaxSteps = 0;

  MobileUseAgent(
    DeviceControlService deviceControl,
    MobileUseAiService aiService,
    MobileUseActionHandler actionHandler,
  ) : _executor = TaskExecutor(deviceControl, aiService, actionHandler);

  /// Current task status (for UI polling).
  TaskStatus get currentStatus => _currentStatus;
  String get currentMessage => _currentMessage;
  int get currentStep => _currentStep;
  int get currentMaxSteps => _currentMaxSteps;
  bool get isRunning =>
      _currentStatus == TaskStatus.planning ||
      _currentStatus == TaskStatus.executing ||
      _currentStatus == TaskStatus.verifying ||
      _currentStatus == TaskStatus.waitingForConfirmation;

  /// Request cancellation of the running task.
  void cancel() => _cancelRequested = true;

  /// Validates a structured task before execution. Returns an error message
  /// if invalid, or null if acceptable.
  String? validateTask(MobileTask task) {
    if (task.id.isEmpty) return 'Task is missing an id.';
    if (task.goal.trim().isEmpty) return 'Task is missing a goal.';
    if (task.maxSteps <= 0) return 'Task maxSteps must be greater than zero.';
    if (task.allowedActions.isEmpty) return 'Task has no allowed actions.';
    final overlap = task.allowedActions
        .where((a) => task.blockedActions.contains(a))
        .toList();
    if (overlap.isNotEmpty) {
      return 'Actions are both allowed and blocked: ${overlap.join(", ")}';
    }
    return null;
  }

  /// Execute a structured task end-to-end.
  ///
  /// Emits progress via the returned stream and returns the final
  /// [TaskResult]. The caller (Beatrice) is responsible for checking
  /// `requiresConfirmation` and obtaining user consent before calling this
  /// for high-risk tasks.
  Future<TaskResult> runTask(MobileTask task) async {
    // 1. Validate.
    final validationError = validateTask(task);
    if (validationError != null) {
      return TaskResult(
        taskId: task.id,
        success: false,
        verificationStatus: VerificationStatus.failed,
        resultSummary: 'The task was rejected before execution.',
        importantObservations: const [],
        actionsPerformed: const [],
        failureReason: validationError,
        stepsTaken: 0,
        cancelled: false,
      );
    }

    // 2. Reset state.
    _cancelRequested = false;
    _currentStatus = TaskStatus.planning;
    _currentMessage = "I'm figuring out how to do this.";
    _currentStep = 0;
    _currentMaxSteps = task.maxSteps;

    // 3. Run the executor loop.
    final result = await _executor.execute(
      task,
      onProgress: (message, step) {
        _currentStatus = TaskStatus.executing;
        _currentMessage = message;
        _currentStep = step;
      },
      shouldCancel: () {
        if (_cancelRequested) {
          _cancelRequested = false;
          return true;
        }
        return false;
      },
    );

    // 4. Handle cancellation.
    if (result.cancelled) {
      _currentStatus = TaskStatus.cancelled;
      _currentMessage = 'The task was cancelled.';
      return TaskResult(
        taskId: task.id,
        success: false,
        verificationStatus: VerificationStatus.unverified,
        resultSummary: 'The task was cancelled.',
        importantObservations: result.verifiedObservations,
        actionsPerformed: result.actions,
        failureReason: 'User cancelled the task while it was running.',
        stepsTaken: result.stepsTaken,
        cancelled: true,
      );
    }

    // 5. Final verification gate.
    _currentStatus = TaskStatus.verifying;
    _currentMessage = "I'm verifying the result.";

    // 6. Assemble the structured result.
    final success = result.verificationStatus == VerificationStatus.verified;
    final summary = success
        ? (result.verifiedObservations.isNotEmpty
            ? result.verifiedObservations.join(' ')
            : 'Done — I completed the task and verified the result.')
        : (result.failureReason.isNotEmpty
            ? result.failureReason
            : 'I could not reliably complete or verify the task.');

    _currentStatus =
        success ? TaskStatus.completed : TaskStatus.failed;
    _currentMessage = summary;

    return TaskResult(
      taskId: task.id,
      success: success,
      verificationStatus: result.verificationStatus,
      resultSummary: summary,
      importantObservations: result.verifiedObservations,
      actionsPerformed: result.actions,
      failureReason: success ? null : result.failureReason,
      stepsTaken: result.stepsTaken,
      cancelled: false,
    );
  }

  /// Reset to idle.
  void reset() {
    _cancelRequested = false;
    _currentStatus = TaskStatus.planning;
    _currentMessage = '';
    _currentStep = 0;
    _currentMaxSteps = 0;
  }
}

File: lib/private-agent/orchestrator.ts
/**
 * PrivateAgent — Orchestrator (MobileUseAgent interface)
 *
 * This is the public entry point Beatrice uses to talk to PrivateAgent.
 * It receives a structured task, validates it, runs the executor loop,
 * streams progress events to the UI store, honours cancellation, and
 * returns a structured result.
 *
 * The orchestrator is the single place that enforces:
 *   - Task validation before execution.
 *   - Rejection of any action not in the allowed-action list.
 *   - Max-step enforcement.
 *   - Confirmation gating for high-risk tasks.
 *   - Structured result assembly.
 *
 * @license SPDX-License-Identifier: Apache-2.0
 */

import { executeTask } from './executor';
import { usePrivateAgent } from './store';
import { verifyFinalOutcome } from './verifier';
import type { LlmConfig } from './llm-client';
import type {
  CompletionStatus,
  ProgressEvent,
  StructuredResult,
  StructuredTask,
  TaskState,
  VerificationStatus,
} from './types';

export interface RunOptions {
  apiKey: string;
  baseUrl: string;
  model?: string;
}

/**
 * Validates a structured task before execution. Returns an error message
 * if the task is invalid, or null if it is acceptable.
 */
export function validateTask(task: StructuredTask): string | null {
  if (!task.taskId) return 'Task is missing a task id.';
  if (!task.goal || task.goal.trim().length === 0) return 'Task is missing a goal.';
  if (task.maxSteps <= 0) return 'Task maxSteps must be greater than zero.';
  if (task.allowedActions.length === 0) return 'Task has no allowed actions.';

  // An action cannot be both allowed and blocked.
  const overlap = task.allowedActions.filter(a =>
    task.blockedActions.includes(a),
  );
  if (overlap.length > 0) {
    return `Actions are both allowed and blocked: ${overlap.join(', ')}`;
  }

  return null;
}

/**
 * Executes a structured task via PrivateAgent.
 *
 * Emits progress events to the PrivateAgent store as it runs and returns
 * the final structured result. The caller (Beatrice) is responsible for
 * checking `requiresConfirmation` on the task and obtaining user consent
 * before calling this function for high-risk tasks.
 */
export async function runStructuredTask(
  task: StructuredTask,
  options: RunOptions,
): Promise<StructuredResult> {
  const store = usePrivateAgent.getState();
  const { apiKey, baseUrl, model } = options;
  const llm: LlmConfig = { apiKey, baseUrl, model: model ?? 'eburon-code-fast:latest' };

  // 1. Validate the task before execution.
  const validationError = validateTask(task);
  if (validationError) {
    const result: StructuredResult = {
      taskId: task.taskId,
      completionStatus: 'failure',
      verificationStatus: 'failed',
      resultSummary: 'The task was rejected before execution.',
      importantObservations: [],
      actionsPerformed: [],
      failureReason: validationError,
      stepsTaken: 0,
    };
    store.completeTask(result);
    return result;
  }

  // 2. Mark the task as started in the UI store.
  store.startTask(task.taskId, task.maxSteps);

  // 3. Emit a planning progress event.
  emitProgress(task.taskId, 'planning', 'I\'m figuring out how to do this.', 0, task.maxSteps);

  // 4. Run the executor loop.
  const executorResult = await executeTask(task, {
    apiKey,
    baseUrl,
    model,
    onProgress: (message, stepNumber) => {
      const state: TaskState = stepNumber === 0 ? 'executing' : 'executing';
      emitProgress(task.taskId, state, message, stepNumber, task.maxSteps);
    },
    shouldCancel: () => usePrivateAgent.getState().consumeCancel(),
  });

  // 5. If cancelled, return a cancelled result.
  if (executorResult.cancelled) {
    const result: StructuredResult = {
      taskId: task.taskId,
      completionStatus: 'cancelled',
      verificationStatus: 'unverified',
      resultSummary: 'The task was cancelled.',
      importantObservations: executorResult.verifiedObservations,
      actionsPerformed: executorResult.actions,
      failureReason: 'User cancelled the task while it was running.',
      stepsTaken: executorResult.stepsTaken,
    };
    store.completeTask(result);
    return result;
  }

  // 6. Final verification gate.
  emitProgress(task.taskId, 'verifying', 'I\'m verifying the result.', executorResult.stepsTaken, task.maxSteps);

  let finalVerificationStatus: VerificationStatus = executorResult.verificationStatus;
  let finalObservations = executorResult.verifiedObservations;
  let finalSummary = '';

  if (executorResult.finalScreen && finalVerificationStatus !== 'verified') {
    // One more verification pass against the final screen.
    const finalCheck = await verifyFinalOutcome(
      executorResult.finalScreen,
      task.goal,
      llm,
    );
    finalVerificationStatus = finalCheck.status;
    if (finalCheck.observations.length > 0) {
      finalObservations = finalCheck.observations;
    }
  }

  // 7. Assemble the structured result.
  const completionStatus: CompletionStatus =
    finalVerificationStatus === 'verified'
      ? 'success'
      : 'failure';

  if (completionStatus === 'success') {
    finalSummary =
      finalObservations.length > 0
        ? finalObservations.join(' ')
        : 'Done — I completed the task and verified the result.';
  } else {
    finalSummary =
      executorResult.failureReason ??
      'I could not reliably complete or verify the task.';
  }

  const result: StructuredResult = {
    taskId: task.taskId,
    completionStatus,
    verificationStatus: finalVerificationStatus,
    resultSummary: finalSummary,
    importantObservations: finalObservations,
    actionsPerformed: executorResult.actions,
    failureReason: completionStatus === 'success' ? null : (executorResult.failureReason ?? finalSummary),
    stepsTaken: executorResult.stepsTaken,
  };

  // 8. Mark complete in the store.
  store.completeTask(result);
  return result;
}

/**
 * Emits a progress event to both the store and any external listeners.
 */
function emitProgress(
  taskId: string,
  state: TaskState,
  message: string,
  stepNumber: number,
  maxSteps: number,
): void {
  const event: ProgressEvent = {
    taskId,
    state,
    message,
    stepNumber,
    maxSteps,
    timestamp: new Date().toISOString(),
  };
  usePrivateAgent.getState().emitProgress(event);
}

/**
 * Requests cancellation of the currently running task.
 * The executor polls this between steps and will abort at the next safe point.
 */
export function cancelRunningTask(): void {
  usePrivateAgent.getState().requestCancel();
}

/**
 * Resets the PrivateAgent store to idle. Call this after Beatrice has
 * consumed the final result and is ready for the next task.
 */
export function resetPrivateAgent(): void {
  usePrivateAgent.getState().reset();
}